vtm/src/org/oscim/layers/tile/buildings/BuildingLayer.java

```diff
@@ -77,63 +106,126 @@ public boolean process(MapTile tile, RenderBuckets buckets, MapElement element,
 
         ExtrusionStyle extrusion = (ExtrusionStyle) style.current();
 
+        // Filter all building elements
+        // TODO #TagFromTheme: load from theme or decode tags to generalize mapsforge tags
+        boolean isBuildingPart = element.tags.containsKey(Tag.KEY_BUILDING_PART)
+                || (element.tags.containsKey("kind") && element.tags.getValue("kind").equals("building_part"));
+        if (element.tags.containsKey(Tag.KEY_BUILDING) || isBuildingPart
+                || (element.tags.containsKey("kind") && element.tags.getValue("kind").equals("building"))
+                ) {
+            List<BuildingElement> buildingElements;
+            if (mBuildings.containsKey(tile.hashCode())) {
+                buildingElements = mBuildings.get(tile.hashCode());
+            } else {
+                buildingElements = new ArrayList<>();
+                mBuildings.put(tile.hashCode(), buildingElements);
+            }
+            element = element.clone(); // Deep copy, because element will be cleared
```

Comment:
Not sure how much this cloning affects memory / performance?

Reply:
Yes, was my worry, too. But i didn't notice any lag on my devices. Before, I tried to introduce a new MapElement for each _element_ in the different `TileDecoder`s but that would cause maybe more traffic. And there aren't that much parts in OSM so far.